### ISSUE TITLE
VERIFY3(0 == dmu_object_dirty_raw(os, object, tx)) failed (0 == 17 during RAW zfs send  (= encrypted zfs send)

### ISSUE THREAD

**Issue by ronnyegner** [zfsonlinux/zfs#6844]
### System information SOURCE

Distribution Name       |  Ubuntu
  ---                                  |     --- 
Distribution Version    |  17.04
Linux Kernel                 |  4.10.0-24-generic
Architecture                 | x86_64
ZFS Version                  | v0.7.0-128_gc616dcf8b
SPL Version                  | 0.7.0-16_g0cefc9d

### System information TARGET

Distribution Name       |  Debian
  ---                                  |     --- 
Distribution Version    |  9.1
Linux Kernel                 |  4.9.0-3-amd64
Architecture                 | x86_64
ZFS Version                  | v0.7.0-128_gc616dcf8b
SPL Version                  | v0.7.0-16_g0cefc9d

### Describe the problem you're observing
Hi,
i am doing a incremental ZFS raw send and i get the following error on the ZFS receiving side. Both systems are running almost up-to-date ZFS master with the latest ZFS encryption patches.

### Describe how to reproduce the problem
Issue occurs during zfs send from source to TARGET:

` zfs  send -cL -w -i transfer_1 pool2/encrypt/data/Privat@transfer_2 | pv -L 700k  | ssh TARGET zfs recv -s backup/Privat`

The issue can be reproduced at will. 

### Include any warning/errors/backtraces from the system logs

`BACKUP` hostname = TARGET System!

```
Message from syslogd@backup at Nov  7 12:14:27 ...
 kernel:[51973.778486] VERIFY3(0 == dmu_object_dirty_raw(os, object, tx)) failed (0 == 17)

Message from syslogd@backup at Nov  7 12:14:27 ...
 kernel:[51973.778574] PANIC at dmu.c:937:dmu_free_long_object_impl()

Nov  7 12:14:27 backup kernel: [51973.778486] VERIFY3(0 == dmu_object_dirty_raw(os, object, tx)) failed (0 == 17)
Nov  7 12:14:27 backup kernel: [51973.778574] PANIC at dmu.c:937:dmu_free_long_object_impl()
Nov  7 12:14:27 backup kernel: [51973.778623] Showing stack for process 16539
Nov  7 12:14:27 backup kernel: [51973.778672] CPU: 4 PID: 16539 Comm: receive_writer Tainted: P           O    4.9.0-3-amd64 #1 Debian 4.9.30-2+d
eb9u5
Nov  7 12:14:27 backup kernel: [51973.778767] Hardware name: FUJITSU D3401-H1/D3401-H1, BIOS V5.0.0.11 R1.21.0 for D3401-H1x                    0
5/15/2017
Nov  7 12:14:27 backup kernel: [51973.778864]  0000000000000000 ffffffffa53285b4 ffffffffc0fb6104 ffffac12098cbd10
Nov  7 12:14:27 backup kernel: [51973.778955]  ffffffffc08be385 ffffffffc0e4cd88 ffffffff00000030 ffffac12098cbd20
Nov  7 12:14:27 backup kernel: [51973.779044]  ffffac12098cbcc0 2833594649524556 756d64203d3d2030 5f7463656a626f5f
Nov  7 12:14:27 backup kernel: [51973.779133] Call Trace:
Nov  7 12:14:27 backup kernel: [51973.779179]  [<ffffffffa53285b4>] ? dump_stack+0x5c/0x78
Nov  7 12:14:27 backup kernel: [51973.779244]  [<ffffffffc08be385>] ? spl_panic+0xc5/0x100 [spl]
Nov  7 12:14:27 backup kernel: [51973.779360]  [<ffffffffc0e4cd88>] ? dbuf_read+0x738/0xb40 [zfs]
Nov  7 12:14:27 backup kernel: [51973.779485]  [<ffffffffc0e6e377>] ? dnode_hold_impl+0xb77/0xc60 [zfs]
Nov  7 12:14:27 backup kernel: [51973.779541]  [<ffffffffa560390e>] ? mutex_lock+0xe/0x30
Nov  7 12:14:27 backup kernel: [51973.779678]  [<ffffffffc0e9c86f>] ? multilist_sublist_lock_obj+0x1f/0x40 [zfs]
Nov  7 12:14:27 backup kernel: [51973.779826]  [<ffffffffc0e598e4>] ? dmu_free_long_object_impl.part.11+0xb4/0xe0 [zfs]
Nov  7 12:14:27 backup kernel: [51973.779976]  [<ffffffffc0e65f64>] ? receive_freeobjects.isra.11+0x54/0x110 [zfs]
Nov  7 12:14:27 backup kernel: [51973.780125]  [<ffffffffc0e663ba>] ? receive_writer_thread+0x39a/0xd20 [zfs]
Nov  7 12:14:27 backup kernel: [51973.780190]  [<ffffffffc08bb292>] ? thread_generic_wrapper+0x62/0x80 [spl]
Nov  7 12:14:27 backup kernel: [51973.780309]  [<ffffffffc0e66020>] ? receive_freeobjects.isra.11+0x110/0x110 [zfs]
Nov  7 12:14:27 backup kernel: [51973.780401]  [<ffffffffc08bb230>] ? __thread_exit+0x20/0x20 [spl]
Nov  7 12:14:27 backup kernel: [51973.780460]  [<ffffffffc08bb29d>] ? thread_generic_wrapper+0x6d/0x80 [spl]
Nov  7 12:14:27 backup kernel: [51973.780519]  [<ffffffffa50965d7>] ? kthread+0xd7/0xf0
Nov  7 12:14:27 backup kernel: [51973.780570]  [<ffffffffa5096500>] ? kthread_park+0x60/0x60
Nov  7 12:14:27 backup kernel: [51973.780623]  [<ffffffffa56065f5>] ? ret_from_fork+0x25/0x30

```

Subsequent tries to send the stream fail with:

```
root@homenas:~# zfs  send -cL -w -i transfer_1 pool2/encrypt/data/Privat@transfer_2 | pv -L 700k  | ssh TARGET zfs recv -s backup/Privat
cannot receive incremental stream: dataset is busy                                                                              ]
```

I suspect the error is related to the zfs encryption patches since i am doing a zfs raw send, hence i am calling @tcaputi

**Comment by tcaputi** [comment]
This is already fixed in master: 35df0bb5565c81ba52de430108b37eced9de0606

**Comment by ronnyegner** [comment]
Oops. Sorry. missed that.

**Comment by ronnyegner** [comment]
Problem is solved with master 35df0bb. Thanks again Tom.
